### PR TITLE
Standardized Markdown using Prettier

### LINKED ISSUE #1121: Standardize on prettier for formatting of our markdown
Rather than bike shedding formatting suggestions, just let a tool do the work. I suggest using [prettier](https://prettier.io).

* [ ] Add prettier to the project
* [ ] Format everything using prettier
* [ ] Add pre-commit hook to ensure changes conform to prettier

### PR DESCRIPTION
**Pull request by ravireddy07**
This Pull Request resolved the issue, fixes #1121  

--All Markdown files has been formatted using Prettier.

### **Prettier settings:**

- Arrow parens(always)
- Bracket Spacing(True)
- Disable Languages(none-based on your choice)
- End of Line(Auto)
- Insert Pragma(False)
- Jsx Bracket Same Line(False)
- Jsx Single Quote(False)
- Fir code within this line limit(80)
- Prose width(preserve)
- Quote Props(as-needed)
- Require Config(False)
- Require Pragma(False)
- Resolve Global Modules(False)
- Semi(True)
- Single Quote(True)
- Tab Width(2)
- Trailing Comma(none)
- Use Tabs(False)
- Vue Indent Script And Style(False)

### CODE REVIEW

**Comment by Kundan28** [comment]
@ravireddy07 Please link the issue to PR like [this](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)

**Comment by Kundan28** [comment]
And please resolve the conflicts.

**Comment by ravireddy07** [comment]
Conflicts resolved and linked the issue in requested format. Please review @alodahl

**Comment by Kundan28** [comment]
> Conflicts resolved and linked the issue in requested format. Please review @alodahl

Conflicts still shows.

**Comment by Kundan28** [comment]
Nice work @ravireddy07

**Comment by vegetabill** [comment]
@ravireddy07 the changes look good but the important thing for a repository is to ensure that we are all using the same settings and they are enforced.

- Commit the settings you mentioned...isn't it .prettierrc or something?

**Comment by ravireddy07** [comment]
### **Prettier settings:**

- Arrow parens(always)
- Bracket Spacing(True)
- Disable Languages(none-based on your choice)
- End of Line(Auto)
- Insert Pragma(False)
- Jsx Bracket Same Line(False)
- Jsx Single Quote(False)
- Fir code within this line limit(80)
- Prose width(preserve)
- Quote Props(as-needed)
- Require Config(False)
- Require Pragma(False)
- Resolve Global Modules(False)
- Semi(True)
- Single Quote(True)
- Tab Width(2)
- Trailing Comma(none)
- Use Tabs(False)
- Vue Indent Script And Style(False)

**Comment by ravireddy07** [comment]
@alodahl  @vegetabill 
a new markdown file added which consists of prettier settings used for all markdown's in the repo. Please once check.

File: prettier-settings.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# **Prettier Settings**
```
Comment (alodahl) [inline prettier-settings.md]: I think we are not understanding each other.  You have only completed one of the ACs - you've formatted everything:
image

For the other two, you need to *install* prettier into our project - that requires a config file or an eslintrc or something to that effect. Also, the issue requests a pre-commit hook for automatic lint checking, which I would expect to see in the package.json.

**Comment by ravireddy07** [comment]
@alodahl  Added a _.prettierrc_ config file along with a pre-commit hook. please once check now.

**Comment by ravireddy07** [comment]
@Kundan28  rechecked all files and corrected.
@alodahl @vegetabill  added pre-commit hook and ._prettierrc_ config file too. Please check.

**Comment by ravireddy07** [comment]
@Kundan28  i only made all changes till now from the time i got assigned.

**Comment by Kundan28** [comment]
> @Kundan28 which changes...?
> i only made all changes till now from the time i got assigned in this PR.

I saw some changes that were made earlier too by you, that's why I asked.

**Comment by ravireddy07** [comment]
Yup, this commits/changes are of resolving of conflicts.

File: package.json [hunk 2]
```diff
@@ -13,15 +13,46 @@
...
+  "dependencies": {
+    "git-format-staged": "^2.1.0",
+    "husky": "^4.2.3",
+    "mrm": "^2.1.1",
+    "precise-commits": "^1.0.2",
+    "pretty-quick": "^2.0.1"
+  },
+  "husky": {
+    "hooks": {
+      "pre-commit": "npm run prettier",
```
Comment (alodahl) [inline package.json]: Having 2 with the same alias is a problem.  Do you need both?  I would recommend something like:
```suggestion
      "pre-commit": "npm run prettier && lint-staged",
```

**Comment by ravireddy07** [comment]
When contributors will commit there changes on their respective branch, At the time of commit, this pre-commit will format the indentation of all **markdown** files which they would like to commit. Explicitly, we don't need to do format anything in markdowns, just write the content and commit it. That pre-commit will format the committed markdown files according to the settings provided in _.prettierrc_ file.

**Comment by alodahl** [comment]
Thanks @ravireddy07, please resolve your conflicts now.

**Comment by Kundan28** [comment]
@ravireddy07 You could have a look at [this](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/resolving-a-merge-conflict-using-the-command-line) for resolving conflicts.

**Comment by ravireddy07** [comment]
Thanks @Kundan28 . All conflicts resolved.
Whenever, new PR(which has markdown) got merged into the Repo. This conflicts are arising.

@alodahl  @vegetabill @gsong  Please check.

**Comment by alodahl** [comment]
@vegetabill can you validate this?  I'm not sure how.

**Comment by Kundan28** [comment]
> @vegetabill can you validate this? I'm not sure how.

Same here.

**Comment by ravireddy07** [comment]
@vegetabill  Sir, please once review it..!

**Comment by alodahl** [comment]
@ravireddy07 how do we validate?

**Comment by ravireddy07** [comment]
@alodahl yup, sure. I just asked to review/check it either it meeting your expectations or not...!!!

**Comment by alodahl** [comment]
I am confused. Tell us how to validate that this works.

**Comment by ravireddy07** [comment]
Don't install prettier on your editor. Make changes in a file and commit it. And see the format of that file after committing whether it's following/working according to the _.prettierrc_ configuration.

**Comment by alodahl** [comment]
Can you give me an example? Like, should I try to make an incorrect indentation, commit, and then see that the actual commit changes have a corrected indentation?

**Comment by ravireddy07** [comment]
Yes Exactly,
Take any file from [master](https://github.com/Techtonica/curriculum), because none of the markdown files have indentation as prettier. rename it something else name so that we can delete it later. So, take any file from master and commit it on my branch(because, pre-commit and configuration files are on my branch.) And after commiting, check that file indentation with the file which has correct indentation. You can find the correct indentation file on this current PR.

**Comment by ravireddy07** [comment]
> Add a line or two in CONTRIBUTING.md about the fact that prettier is installed, and what it will affect.
> 
> Also, we just merged a bunch of stuff, so you have more conflicts to fix. Sorry 😁

Solved...

**Comment by alodahl** [comment]
I'm not seeing any changes - why would it let me do this? THe funny thing is that ctrl+opt+F works  to reformat now!  But I'm not seeing anything happen on commit.
image

**Comment by alodahl** [comment]
Oh wait!  It works if you are using the command line.  Awesome.  I was using my Atom git panel.

**Comment by Kundan28** [comment]
Updated the scoresheet too.

File: package.json [hunk 3]
```diff
@@ -13,15 +13,45 @@
     "url": "https://github.com/Techtonica/curriculum/issues"
   },
   "scripts": {
-    "validate-links": "remark -u validate-links ."
```
Comment (gsong) [inline package.json:16]: So…why was this removed? Why wouldn't we incorporate this into the pre-commit hook as well?

File: package.json [hunk 4]
```diff
@@ -13,15 +13,45 @@
...
     "remark-validate-links": "^9.0.1"
   },
   "remarkConfig": {
-    "plugins": ["remark-lint-no-dead-urls"]
-  }
+    "plugins": [
+      "remark-lint-no-dead-urls"
+    ]
+  },
+  "dependencies": {
```
Comment (gsong) [inline package.json:34]: Why are these dependencies needed? I don't think any of them are used except for husky, which is already in `devDependencies`.

File: package.json [hunk 5]
```diff
@@ -13,15 +13,45 @@
...
+      "post-commit": "git update-index --again"
+    }
+  },
+  "lint-staged": {
+    "*.md": [
+      "prettier --write",
+      "git add"
+    ]
+  },
+  "pre-commit": [
```
Comment (gsong) [inline package.json:53]: What's this for?

File: package.json [hunk 6]
```diff
@@ -13,15 +13,45 @@
...
+  "husky": {
+    "hooks": {
+      "pre-commit": "npm run prettier && lint-staged",
+      "post-commit": "git update-index --again"
+    }
+  },
+  "lint-staged": {
+    "*.md": [
+      "prettier --write",
+      "git add"
```
Comment (gsong) [inline package.json:50]: This isn't necessary.

**Comment by gsong** [comment]
I should probably just open a PR to fix some of the issues from this PR.

**Comment by alodahl** [comment]
@gsong This was done by someone new to prettier. Please remove anything you find unnecessary if you have the time!  

Regarding pre-commit hook, can you first help me understand what happens automatically and what doesn't?  It's been a while, but I remember it auto-correcting when I commit from the command line.
